launch_testing/launch_testing/markers.py

```diff
@@ -28,14 +30,18 @@ def retry_on_failure(*, times):
     assert times > 0
 
     def _decorator(func):
+        assert 'self' == list(inspect.signature(func).parameters)[0]
         @functools.wraps(func)
-        def _wrapper(*args, **kwargs):
+        def _wrapper(self, *args, **kwargs):
             n = times
             while n > 1:
                 try:
-                    return func(*args, **kwargs)
+                    ret = func(self, *args, **kwargs)
+                    if isinstance(self, unittest.TestCase):
+                        assert self._outcome.success
```

Comment:
Why is this necessary?

Reply:
[`unittest.TestCase.subTest`](https://github.com/python/cpython/blob/master/Lib/unittest/case.py#L547) catches all `AssertionError` exceptions thrown and simply updates the test case result. That allows it to continue going over other sub tests, but breaks this decorator (which expects an `AssertionError` if there's a failure). This extra assertion makes sure that those updates do not pass by unnoticed.